### PR TITLE
Hotkey feature: find hotkeys by heat

### PR DESCRIPTION
**Pull request by littlegreyzhang**
Signed-off-by: littlegreyzhang <826999015@qq.com>

Commit Message: Hotkey feature: find hotkeys by heat

Additional Description:
Add a new feature to find hotkeys
#16096

Risk Level: Low
Testing: New and existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 1]
```diff
@@ -207,6 +207,34 @@ message RedisProxy {
     repeated string commands = 4;
   }
 
+  message FeatureConfig {
+    // [#next-free-field: 6]
+    message HotKey {
+      enum CacheType { LFU = 0; }
```
Comment (lizan) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:213]: add comment for each enum and fields please.

**Review by lizan — CHANGES_REQUESTED** [review]

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto (hunk 1 above)
Comment (littlegreyzhang) [reply api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto]: done

**Comment by littlegreyzhang** [comment]
How do I solve the compilation problem of [fatal error: cannot execute ‘/usr/lib/gcc/x86_64-linux-gnu/9/cc1plus’: execv: Argument list too long](check linux_x64 gcc)?
https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/74866/logs/126

**Comment by lizan** [comment]
/lgtm api

Though I'm not very familiar with Redis itself, @msukalski @henryyyang?

**Comment by littlegreyzhang** [comment]
Please review for me. Thank you.@mattklein123 @msukalski @HenryYYang

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 2]
```diff
@@ -300,6 +364,17 @@ message RedisProxy {
...
+  // Enable some extra features.
+  //
+  // Example:
+  //
+  // .. code-block:: yaml
+  //
+  //    feature_config:
+  //      hotkey: {}
+  //
+  FeatureConfig feature_config = 9;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:377]: From an API perspective I think I would just go ahead and inline `HotKey` here.

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 3]
```diff
@@ -207,6 +207,70 @@ message RedisProxy {
     repeated string commands = 4;
   }
 
+  // FeatureConfig defines the configuration of all extra features.
+  message FeatureConfig {
+    // HotKey defines the configuration of hotkey feature.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:212]: Can you describe more what this feature does? It's not clear from the docs what this means.

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 4]
```diff
@@ -207,6 +207,70 @@ message RedisProxy {
...
+      }];
+
+      // Configure the minimum duration of cache for attenuation operations.
+      // min: 0s
+      // max: 3600s
+      // default: 60s
+      google.protobuf.Duration attenuate_cache_interval = 5 [(validate.rules).duration = {
+        lte {seconds: 3600}
+        gte {}
+      }];
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:255]: Can you describe more what attenuation means? I'm not sure from a quick read.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks just starting with a few high level API comments. Also:
1) This feature should have some arch documentation. I would put it here: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/redis
2) Please also add a release note.

Once the docs and API are set I will do the code review. Thank you!

/wait

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto (hunk 2 above)
Comment (littlegreyzhang) [reply api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:377]: Yes, you are right. But I want to enhance future extensions. Some extra features are all in FeatureConfig, such as HotKey or other.

Comment (mattklein123) [reply api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:377]: If we have future features we can add them directly. I don't think this later of indirection adds any value.

**Comment by mattklein123** [comment]
Merge main?

/wait

**Comment by littlegreyzhang** [comment]
> Merge main?
> 
> /wait

No, I will resolve all conversations and conflicts.

**Comment by littlegreyzhang** [comment]
@mattklein123 What can I do for the error([403 Forbidden - GET https://registry.npmjs.org/wasm2c/-/wasm2c-1.0.0.tgz](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/81231/logs/200))?
I don't think this is my mistake.

**Comment by littlegreyzhang** [comment]
> @mattklein123 What can I do for the error([403 Forbidden - GET https://registry.npmjs.org/wasm2c/-/wasm2c-1.0.0.tgz](https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/81231/logs/200))?
> I don't think this is my mistake.

After synchronizing the main line's new code, a recompile is triggered and the error disappears.😂

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 5]
```diff
@@ -207,6 +207,75 @@ message RedisProxy {
...
+    // default: 5s
+    // min: 0.5s
+    // max: 3600s
+    google.protobuf.Duration collect_dispatch_interval = 3 [(validate.rules).duration = {
+      lte {seconds: 3600}
+      gte {nanos: 500000000}
+    }];
+
+    // Configure the dispatch interval for the attenuation operation.
+    // To avoid full heat over a period of time, we need to attenuate it regularly.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:241]: Can you define "attenuate." I'm still not completely sure what this means. I think more definition in layman's terms followed by the example would help.

File: source/extensions/filters/network/redis_proxy/command_splitter_impl.cc [hunk 6]
```diff
@@ -177,6 +180,16 @@ SplitRequestPtr EvalRequest::create(Router& router, Common::Redis::RespValuePtr&
   std::unique_ptr<EvalRequest> request_ptr{
       new EvalRequest(callbacks, command_stats, time_source, delay_command_latency)};
 
+  try {
+    // maybe throw std::invalid_argument or std::out_of_range
+    uint64_t key_max_offset(3 + std::stoull(incoming_request->asArray()[2].asString()));
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/command_splitter_impl.cc:185]: You can use `absl::SimpleAtoi` to avoid the exception handling here.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, sorry for the review delay. I have some high level doc comments, but this PR is still too large to review. Can you start splitting this part into multiple PRs. I would figure out the right way of splitting it up, probably starting with the cache/hot key implementation, then actually wiring it up into the filter.

/wait

**Comment by littlegreyzhang** [comment]
> Thanks, sorry for the review delay. I have some high level doc comments, but this PR is still too large to review. Can you start splitting this part into multiple PRs. I would figure out the right way of splitting it up, probably starting with the cache/hot key implementation, then actually wiring it up into the filter.
> 
> /wait

Do I have to split this PR? I don't really want to do this because I think it's a complete feature and it's already there. Can I follow your advice next time?

But if you insist, I will do it this time. It will take some time.

😊

**Comment by mattklein123** [comment]
> But if you insist, I will do it this time. It will take some time.

Yes, I insist. This is too much code to review. Please split into some pieces. I don't think it will be very difficult to do that with this PR.

**Comment by littlegreyzhang** [comment]
Split  some code into #17333. @mattklein123
